### DISCUSSION TITLE
Logs WorkFlow

### DISCUSSION THREAD

**Discussion by ncvescera** [CristianCosci/GMRES_python#2] [General]
Logs

**Comment by ncvescera** [comment]
## Fri 2 Dec

- Iniziato sviluppo progetto
- Trovato materiale da studiare

**Comment by CristianCosci** [comment]
## **Mon 5 Dec**
- Trovata libreria per convertire e caricare matrice sparse di grande dimensione da [www.sparse.tamu.edu](https://sparse.tamu.edu/)
- Trovato formato utile in scipy per lavorare con matrici sparse di grande dimensione (csr_matrix)
- Fatta una prova con due matrici molto grandi e simulato un prodotto matrice vettore per avere un'idea delle tempistiche
- Fatto riepilogo delle cose da implementare

**Comment by CristianCosci** [comment]
## **Tue 6 Dec**
- Prime prove di caricamento di matrici di grandi dimensioni sia in python che in Octave (verrà quindi fatta un'implementazione sia in python che in Octave)
- Iniziato progetto e scrittura algoritmo GMRES
- Scritta funzione GMRES ma con alcuni bug da risolvere:
    - Incontrate prime problematiche sul least squares
- Riorganizzata struttura progetto

**Comment by CristianCosci** [comment]
## **Wed 7 Dec**
- Implementata prima versione di GMRES e fatti primi confronti con la versione scipy per valutarne le prestazioni
- Confronto con il professore su come procedere nell'implementazione

**Comment by ncvescera** [comment]
## Thu 8 Dec

- Tentato di capire Ortogonalizzazione (per ora non ci siamo riusciti 🙃) nell'Arnoldi Iteration
- Suddivisi compiti:
    - **Cosci**: ri-implementazione di GMRES in python con successivi test e grafici (pdf bello)
    - **Fagiolo**: implementazione dell'Arnoldi Iteration come funzione, Test su Eighenvalues di H e confronto con quelli di A. Confrontare anche il miglioramento dell'approssiamzione dei eighenvalues di H ad ongi iterazioni di Arnoldi.
    - **Vescera**: impelementazione di GMRES in Octave con successivi test e grafici.

**Comment by ncvescera** [comment]
## Fri 9 Dec

- Implementato GMRES in Octave (Jupyter Notebook)
- Ri-organizzato Notebook ed emoji 🐒 
- Ho trovato un'implementazione simile alla nostra, con la differenza che non aggiorna "in place" `Q` e dovrebbe utilizzare l'ortogonalizzazione.
    - Ho effettuato alcuni test e questa implementazione converge solo se la dimensione della matrice è uguale al numero massimo di iterazioni 🤔  
    - Ho trovato alcuni esempi in cui converge esattamente alla soluzione. Sono riportati nella sezione del Notebook `Test 2 - randnorm()`.

_Il log riguarda i commit a88ebcf, 856a9ec, c4380f0, 5d2d8b2, 15b6215_

**Comment by F-a-b-r-i-z-i-o** [comment]
## **Fri 9 Dec**

- Implementata arnoldi iteration ispirata a wikipedia   
   - Test con matrice 10x10 fatta a mano 
- Trovati Auto-Valori matrice H e confrontati con metodo lilnag.eigh( ) di numpy
   - Auto-Valori trovati su H sembrano approssimare quelli trovati con il metodo, questo mi fa suppore sia corretto?
- Trovata la "discesa" degli auto-valori in base all'iterazione

#### TO DO
- Eliminare print debug codice 
- Fare qualche grafico 
- Capire come viene approssimato auto-vettore generato
- Verificare problemi approssimazione

**Comment by ncvescera** [comment]
## Sat 10 Dec

- Implementato GMRES con Restart 🗿
- Effettuati test su GMRES con Restart: carino, va un po' meglio di quello normale ma comunque non riesce a convergere e dopo troppi restart inizia a divergere 🙃 
- Forse ho trovato come effettuare l'ortogonalizzazione e questo sembra apportare delle migliorie alla convergenza di GMRES 🥇 

_Il log riguarda i commit 45a6af5d0597b14ce1fcd8243f28e1081a910909, 9a4e2659f8ea167f8f488c191f541598c859cfe1_

**Comment by F-a-b-r-i-z-i-o** [comment]
# Sat 10 Dec
- Trovato vettore di autovalori e confrontato con metodo. 
- Plot grafico di differenza tra auto-valore con meotodo, e auto-valore trovato con algoritmo 
- Plot matrici, H,Q
- Implementazione di lanczos e verifica che approssimazione auto-valore migliore di arnoldi
- Plot approssimazione auto-valori ad ogni iterazione 

<hr>

## To do 

<hr>

- **MIGLIORARE IL CODICE**
- Test su lanczos con relativi plot

**Comment by ncvescera** [comment]
## Tue 13 Dec

- Rivista implementazione di GMRES e GMRES(m) (piccole modifiche e commenti).
- Aggiunti test su GMRES(m) con matrici identità, matrici identità con valori sulla diagonale $> 2$, matrici con elementi compresi tra $0$ e $1$ con elementi sulla diagonale $>2$
- Abbiamo migliorie sulla convergenza 📈 

_Il log è relativo ai commit 40defdac6f65ef1f3b8910e95cdec79eb6700e73, 7f02926f68a13b94cae2af54fe2a27c8f04a60c6_

**Comment by F-a-b-r-i-z-i-o** [comment]
# Tue 13 Dec
- Implementazione metodo delle potenze 
- Implementazione funzione per il calcolo dell'errore relativo 
- Verifica convergenza auto-valori metodo delle potenze 
- Plot grafici

**Comment by ncvescera** [comment]
## Thu 15 Dec

- Risolti problemi con le formule latex nel README. (pr #5 )
- Aggiornato Notebook Octave
  - `mgorth(v, Q)` fa convergere `GMRES(m)`, `gramschmidt(v)` no 🙃 

_Il log riguarda la pull request #5 e il commit e5afc20_

**Comment by ncvescera** [comment]
## Fri 16 Dec

- Rivisto e iniziato a sistemare il Notebook Octave:
  - Aggiunti nuovi test per far vedere il processo
    che ci ha condotto alla soluzione
  - Aggiunta una funzione per generare matrici sparse
    con cui GMRES converge
  - Rivisto GMRES con le modifiche del prof
  - Aggiunta una funzione per stampare autovalori
    e la prima parte della matrice passata
  - Migliorati leggermente i plot

- Iniziato refactoring di `GMRES(m)`
   - ci sono problemi con la convergenza. Le nuove modifiche lo hanno rotto.
 
_Il log è relativo al commit 6d4b902_
